test/lib/PlacesTestUtils.js

```diff
@@ -24,22 +25,59 @@ const PlacesTestUtils = Object.freeze({
    *            [optional] visitDate: visit date in microseconds from the epoch
    *            [optional] referrer: nsIURI of the referrer for this visit
    *          }
-   *
-   * @return {Promise}
-   *            resolves When all visits have been added successfully.
-   *            rejects JavaScript exception.
    */
   addVisits: Task.async(function*(placeInfo) {
-    let promise = new Promise((resolve, reject) => {
-      let places = [];
-      if (placeInfo instanceof Ci.nsIURI) {
-        places.push({uri: placeInfo});
-      } else if (Array.isArray(placeInfo)) {
-        places = places.concat(placeInfo);
-      } else {
-        places.push(placeInfo);
+    let places = [];
+
+    if (placeInfo instanceof Ci.nsIURI) {
+      places.push({uri: placeInfo});
+    } else if (Array.isArray(placeInfo)) {
+      places = places.concat(placeInfo);
+    } else {
+      places.push(placeInfo);
+    }
+
+    let urlList = places.map(place => place.uri.spec);
+    let urlSet = new Set(urlList);
+    let urlCount = {};
+    for (let url of urlList) {
+      urlCount[url] = 0;
+    }
+    // there are 3 notifications per place inserted
+    let totalCount = urlList.length * 3;
+    let notifCount = 0;
+
+    let notifPromise = new Promise(resolve => {
```

Comment:
can you maybe rename this to something or add a comment about needing to wait for notifications before doing the insert, it's a bit unclear just by looking at it immediately why it's needed